prow/plugins/blunderbuss/blunderbuss.go

```diff
@@ -334,34 +360,74 @@ func findReviewer(ghc githubClient, log *logrus.Entry, useStatusAvailability boo
 			// we've already verified this reviewer is busy
 			continue
 		}
-		busy, err := isUserBusy(ghc, candidate)
-		if err != nil {
-			log.Errorf("error checking user availability: %v", err)
-		}
-		if !busy {
-			return candidate
-		}
-		// if we haven't returned the candidate, then they must be busy.
-		busyReviewers.Insert(candidate)
+		return candidate
 	}
 	return ""
 }
 
-type githubAvailabilityQuery struct {
-	User struct {
-		Login  githubql.String
-		Status struct {
-			IndicatesLimitedAvailability githubql.Boolean
+// prepareLeafReviewersToQuery prepares reviewers list using info from reviewers client crossed with author set
+func prepareLeafReviewersToQuery(rc reviewersClient, ghc githubClient, files []github.PullRequestChange, authorSet sets.String) []string {
+	ownersSeen := sets.NewString()
+	reviewersToQuery := layeredsets.NewString()
+	for _, file := range files {
+		ownersFile := rc.FindReviewersOwnersForFile(file.Filename)
+		if ownersSeen.Has(ownersFile) {
+			continue
 		}
-	} `graphql:"user(login: $user)"`
+		ownersSeen.Insert(ownersFile)
+		reviewers := layeredsets.NewString(rc.LeafReviewers(file.Filename).List()...).Difference(authorSet)
+		reviewersToQuery = reviewersToQuery.Union(reviewers)
+	}
+	return reviewersToQuery.List()
 }
 
-func isUserBusy(ghc githubClient, user string) (bool, error) {
-	var query githubAvailabilityQuery
-	vars := map[string]interface{}{
-		"user": githubql.String(user),
+type githubUserAvailability struct {
+	Login  string
+	Status struct {
+		IndicatesLimitedAvailability bool
+	}
+}
+
+// graphqlTag helps in formating the graphql tag
+func graphqlTag(value string) reflect.StructTag {
+	return reflect.StructTag(strconv.AppendQuote([]byte("graphql:"), value))
+}
+
+// findBusyReviewers finds all the people that are busy and excluded from the review
+func findBusyReviewers(ghc githubClient, log *logrus.Entry, busyReviewers *sets.String, reviewers []string) {
+	const (
+		// will panic if too many fields with too much data in the reflect struct, this setting is very safe
+		githubUserAvailabilityStructLimit = 300
+	)
+	currentIndex := 0
+	numberOfBatches := int(math.Ceil(float64(len(reviewers)) / githubUserAvailabilityStructLimit))
+	for batch := 0; batch < numberOfBatches; batch++ {
+		var fields []reflect.StructField
+		batchError := false
+		structsAdded := 0
+		for i := currentIndex; i < len(reviewers) && structsAdded < githubUserAvailabilityStructLimit; i++ {
+			fields = append(fields, reflect.StructField{
+				Tag:  graphqlTag(fmt.Sprintf("user%d :user(login: %q)", i, reviewers[i])),
+				Name: fmt.Sprintf("User%d", i),
+				Type: reflect.TypeOf(githubUserAvailability{}),
+			})
+			structsAdded++
+		}
+		query := reflect.New(reflect.StructOf(fields)).Elem()
+		err := ghc.Query(context.Background(), query.Addr().Interface(), nil)
+		if err != nil {
```

Comment:
Previously, in case of a single error, we counted such reviewers as busy for the review. As the errors repeat over a limit, multiple reviewers were counted as busy ones. The same behavior is more or less recreated here. If error, the whole batch goes to the busy set. Let me know if that is ok.